kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/DataTypeParser.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.internal;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import io.delta.kernel.types.*;
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+
+/**
+ * Parses Delta data types based on the
+ * <a href="https://github.com/delta-io/delta/blob/master/PROTOCOL.md#primitive-types">
+ *     serialization rules </a> outlined in the Delta Protocol.
+ */
+public class DataTypeParser {
+
+    private DataTypeParser() {}
+
+    public static StructType parseSchema(JsonNode json) {
+        DataType parsedType = parseDataType(json);
+        if (parsedType instanceof StructType) {
+            return (StructType) parsedType;
+        } else {
+            throw new IllegalArgumentException(
+                String.format("Could not parse %s as a valid StructType", json));
+        }
+    }
+
+    /**
+     * Parses a Delta data type from JSON. Data types can either be serialized as strings (for
+     * primitive types) or as objects (for complex types).
+     *
+     * For example:
+     * <pre>
+     * // Map type field is serialized as:
+     * {
+     *   "name" : "f",
+     *   "type" : {
+     *     "type" : "map",
+     *     "keyType" : "string",
+     *     "valueType" : "string",
+     *     "valueContainsNull" : true
+     *   },
+     *   "nullable" : true,
+     *   "metadata" : { }
+     * }
+     *
+     * // Integer type field serialized as:
+     * {
+     *   "name" : "a",
+     *   "type" : "integer",
+     *   "nullable" : false,
+     *   "metadata" : { }
+     * }
+     * </pre>
+     */
+    public static DataType parseDataType(JsonNode json) {
+        switch (json.getNodeType()) {
+            case STRING:
+                // simple types are stored as just a string
+                return nameToType(json.textValue());
+            case OBJECT:
+                // complex types (array, map, or struct are stored as JSON objects)
+                String type = getStringField(json, "type");
```

Comment:
consider if lowercasing is needed here.

Reply:
I'm thinking we don't _need_ accomodate for this case as it wouldn't be a valid schema but we still could. What do you think?